### DISCUSSION TITLE
Switching between models for automatic segmentation from command line

### DISCUSSION THREAD

**Discussion by lukasvanderstricht** [Project-MONAI/MONAILabel#1494] [Q&A] [answered]
Hi everyone

I am training some modules via MONAILabel and 3D Slicer and was currently I am working with two models that I have trained (and are in the same app). For the workflow I want to achieve, I would want to automate the process as much as possible, and hence I am looking at ways to turn the workflow into a Python script. 

**Here is in broad lines what I want to do:**
1. Run automatic segmentation with my first model, let's call it `deepedit1`
2. Calculate some volumetric properties of the resulting segments
3. Run automatic segmentation (on the same volume), with my second model, let's call it `deepedit2`
4. Calculate some volumetric properties of the resulting segments

**Here is what I have done already:**
I am able to calculate the the volumetric properties with the `segment statistics` module in Python, so steps 2 and 4 are no problem. I also know how to trigger the automatic segmentation, namely with the command `slicer.modules.MONAILabelWIDGET.onClickSegmentation()`.

**Now here is my problem**
I start my MONAILabel server by loading in my two models, something like this:
`monailabel start_server --app radiology/ --studies data/train_images/ --conf models deepedit1,deepedit2`
This ensures that in the Slicer GUI, under the tab _Automatic segmentation_, I can choose which model I want to use for the automatic segmentation.
I however cannot figure out how I could specify the model used for automatic segmentation using just Python commands. I tried giving the model as a parameter for the `onClickSegmentation()` function, but that doesn't work.

**Alternatively:**
I also tried using the MONAILabel Logic module to circumvent this problem. More specifically, I tried something along the following lines:

```
logic = slicer.util.getModuleLogic('MONAILabel')
result_file,params = logic.infer("deepedit1_seg","/home/data/train_images/DICOMX")
```

where "DICOMX" is the name of the file that is currently loaded into Slicer.
When I do this, however, the labels seem to be computed, yet they do not show up in my viewers and also not in my data tree in Slicer. If I run the following command:
`slicer.util.loadSegmentation(result_file)`
the segmentation is visible and loaded into Slicer, but not on the segmentation node where the my segmentation would be if I just clicked to run automatic segmentation in the GUI. On top of that, the segments don't have the correct name.

**Does anyone have any idea how I could solve this problem?**

Thanks in advance!

Kind regards

Lukas

**Comment by diazandr3s** [comment]
Hi @lukasvanderstricht,

Have you tried running MONAI Label main file as a standard Python script? Here you can find an example: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/main.py#L263-L346

Just specify the studies, models and command you want to run (i.e. train, infer): https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/main.py#L289-L291

Hope this helps,

**Reply by lukasvanderstricht** [reply]
Hi @diazandr3s 

Thanks a lot for your reply! I would however want to run this interference from within a Python file that automates a whole intricate workflow and that is run from the Python console in 3D Slicer. So I think this solution is hard to tie in with other actions I want to automate.

I will keep looking for solutions and  keep you updated!

Again thanks for the help!

**Comment by lukasvanderstricht** [comment, marked as answer]
Hi again,

I found a solution to my problem! Using the following command, you can choose the loaded model which you want to use for automatic segmentation:
`slicer.modules.MONAILabelWidget.ui.segmentationModelSelector.setCurrentText()`

As argument you give a string with the name of the preferred model.
For example, 
`slicer.modules.MONAILabelWidget.ui.segmentationModelSelector.setCurrentText("deepedit_seg")`  
switches the selected model to the `deepedit` model.

Thanks for the help!

Lukas
